### PR TITLE
Initial implementation of a base docker image for docker based Go builds

### PR DESCRIPTION
**Pull request by ernestojeda**
Two files added to ci-mangement repository based on a conversation with @JPWKU.

1. docker/golang-base/Dockerfile - Base docker image definition. Build tools such as gcc, make, etc. included in base image to speed up docker based builds. Also included are optional linting and code coverage tools that we use in our build process to generate reports with SonarQube and Jenkins.

2. docker/golang-base/linter-settings.json - Go Meta Linter default settings. Could be overridden in images that inherit from the base.

### CODE REVIEW

**Comment by JPWKU** [comment]
@tsconn23 Don't merge yet...just a first look.

File: docker/golang-base/Dockerfile
Comment (dweomer) [inline docker/golang-base/Dockerfile]: This code looks vestigial, what is it actually doing?

```
  apk add --no-cache --virtual=.build-dependencies &&\
  ln -s locale.h /usr/include/xlocale.h && \
  rm /usr/include/xlocale.h && \
  apk del .build-dependencies
```

**Comment by dweomer** [comment]
It is lamentable that a "buildpack" is needed. @ernestojeda please correct me if I am wrong but this is at least partly in response to a problem that @jamesrgregg had for
edgexfoundry/device-modbus-go#2 where the default arch golang image did not have the same tooling as the arm64. If such is the case, it might be better to mitigate upstream on the official docker images repository.

That being said, I like the addition of gometalinter.

File: docker/golang-base/Dockerfile
Comment (ernestojeda) [reply docker/golang-base/Dockerfile]: Agreed, this can be removed.

**Comment by ernestojeda** [comment]
@dweomer The "buildpack" was discussed in the recent DevOps working group meeting. You can see reasoning why here: [https://wiki.edgexfoundry.org/display/FA/DevOps+Working+Group?preview=/329486/21823845/DevOps_WG_Meeting_12-06-18(1).pdf](https://wiki.edgexfoundry.org/display/FA/DevOps+Working+Group?preview=/329486/21823845/DevOps_WG_Meeting_12-06-18(1).pdf) on Page #2 `4.Build Pattern going forward for golang (In containers or separate containerization)`

**Comment by dweomer** [comment]
> The "buildpack" was discussed in the recent DevOps working group meeting. You can see reasoning why here: https://wiki.edgexfoundry.org/display/FA/DevOps+Working+Group?preview=/329486/21823845/DevOps_WG_Meeting_12-06-18(1).pdf on Page #2 `4.Build Pattern going forward for golang (In containers or separate containerization)`

Right. But the official `golang` image worked for one arch `amd64` but not for another `arm64` correct? If so, this speaks to an upstream discrepancy. It can be rectified here, in the "downstream", to be sure, but if all of the tooling that one actually needs is/should-be in the upstream then you can have one less thing to manage. This point is largely mooted by the installation of gometalinter, which is great IMO but I would still argue that upstream consistency in tooling across runtime architectures is a must have.

**Comment by JPWKU** [comment]
@ernestojeda I'm working on the jenkins bits to get this hooked into CI.

**Review by JPWKU — COMMENTED** [review]
This LGTM, I just want to get the CI jobs hooked up first that I have in PR here https://github.com/edgexfoundry/ci-management/pull/262

**Comment by JPWKU** [comment]
@ernestojeda DCO will require you to do a `--signoff` on that last commit, you can do a simple amend and signoff

**Comment by JPWKU** [comment]
Also it needs a rebase against upstream/master

**Comment by ernestojeda** [comment]
Ok, I did a merge commit from upstream/master is that OK?

**Comment by JPWKU** [comment]
recheck

**Comment by JPWKU** [comment]
@ernestojeda We got past the last build error, but now it looks like it can't find the linting file.
```19:09:46 Step 6/9 : COPY linter-settings.json /
19:09:47 COPY failed: stat /var/lib/docker/tmp/docker-builder338294390/linter-settings.json: no such file or directory
19:09:47 Build step 'Execute shell' marked build as failure
19:09:47 $ ssh-agent -k
```

**Comment by ernestojeda** [comment]
My latest commit fixed the issue.

**Comment by jamesrgregg** [comment]
recheck

**Review by JPWKU — APPROVED** [review]
LGTM

**Pull request merged**
